test/integration/int_server.cc

```diff
@@ -0,0 +1,815 @@
+}
+
+void ServerConnection::onGoAway() {
+  ENVOY_LOG(warn, "ServerConnection({}) got go away", name_);
+  // TODO how should this be handled? I've never seen it fire.
+}
+
+ServerFilterChain::ServerFilterChain(
+    Envoy::Network::TransportSocketFactory &transport_socket_factory)
+    : transport_socket_factory_(transport_socket_factory) {}
+
+ServerFilterChain::~ServerFilterChain() {}
+
+const Envoy::Network::TransportSocketFactory &
+ServerFilterChain::transportSocketFactory() const {
+  return transport_socket_factory_;
+}
+
+const std::vector<Envoy::Network::FilterFactoryCb>
+    &ServerFilterChain::networkFilterFactories() const {
+  return network_filter_factories_;
+}
+
+LocalListenSocket::LocalListenSocket(
+    Envoy::Network::Address::IpVersion ip_version, uint16_t port,
+    const Envoy::Network::Socket::OptionsSharedPtr &options, bool bind_to_port)
+    : NetworkListenSocket(
+          Envoy::Network::Utility::parseInternetAddress(
+              Envoy::Network::Test::getAnyAddressUrlString(ip_version), port),
+          options, bind_to_port) {}
+
+LocalListenSocket::~LocalListenSocket() {}
+
+ServerCallbackHelper::ServerCallbackHelper(
+    ServerRequestCallback request_callback,
+    ServerAcceptCallback accept_callback, ServerCloseCallback close_callback)
+    : accept_callback_(accept_callback),
+      request_callback_(request_callback),
+      close_callback_(close_callback) {
+  if (request_callback) {
+    request_callback_ = [this, &request_callback](
+                            ServerConnection &connection, ServerStream &stream,
+                            Envoy::Http::HeaderMapPtr request_headers) {
+      ++requests_received_;
+      request_callback(connection, stream, std::move(request_headers));
+    };
+  } else {
+    request_callback_ = [this](ServerConnection &, ServerStream &stream,
+                               Envoy::Http::HeaderMapPtr &&) {
+      ++requests_received_;
+      Envoy::Http::TestHeaderMapImpl response{{":status", "200"}};
+      stream.sendResponseHeaders(response);
+    };
+  }
+
+  if (accept_callback) {
+    accept_callback_ =
+        [this, &accept_callback](
+            ServerConnection &connection) -> ServerCallbackResult {
+      ++accepts_;
+      return accept_callback(connection);
+    };
+  } else {
+    accept_callback_ = [this](ServerConnection &) -> ServerCallbackResult {
+      ++accepts_;
+      return ServerCallbackResult::CONTINUE;
+    };
+  }
+
+  if (close_callback) {
+    close_callback_ = [this, &close_callback](ServerConnection &connection,
+                                              ServerCloseReason reason) {
+      switch (reason) {
+        case ServerCloseReason::REMOTE_CLOSE:
+          ++remote_closes_;
+          break;
+        case ServerCloseReason::LOCAL_CLOSE:
+          ++local_closes_;
+          break;
+      }
+
+      close_callback(connection, reason);
+      std::unique_lock<std::mutex> lock(mutex_);
+      condvar_.notify_one();
```

Comment:
@jplevyak if I use absl::Mutext instead of the mutex+condvar pair, this block becomes:

```
        absl::MutexLock lock(&mutex_);
        switch (reason) {
          case ServerCloseReason::REMOTE_CLOSE:
            ++remote_closes_;
            break;
          case ServerCloseReason::LOCAL_CLOSE:
            ++local_closes_;
            break;
        }
      }
 
      close_callback(connection, reason);
```

It may look a bit cleaner, but now close_callback is called under lock when before it wasn't.  Alternatively I could move the counter increments after the close_callback is called, but that too is new behavior.  And I definitely don't want to call release the lock / do the equivalent of the notification before the callback.  That would make the corresponding wait() return prematurely.

My head is far enough out of this code that I'd rather not introduce a substantive change with the 1.1 deadline looming.  Can I have a pass?  Would be happy to revisit post 1.1.

Reply:
I think you can just scope it, so that lock is going to be released before the callback, i.e
```
{
    absl::MutexLock lock(&mutex_);
    
    switch (reason) {
      case ServerCloseReason::REMOTE_CLOSE:
        ++remote_closes_;
        break;
      case ServerCloseReason::LOCAL_CLOSE:
        ++local_closes_;
        break;
    }
}

close_callback(connection, reason);
```

Reply:
But I the lock going out of scope combines both an unlock and a notify/signal.  That signal is now sent before the close_callback which will introduce a race condition.  The close_callback() may still be doing something when unit test thinks everything has finished.... so the counters will be off and the test may flake

Reply:
Read this:  https://abseil.io/about/design/mutex

Reply:
+1 for @PiotrSikora 
@duderino The current code doesn't guard you from
```close_callback() may still be doing something when unit test thinks everything has finished```
A sequence could be 
1. Thread 1: ++remote_close  // suppose this is the last one
1. Thread 2: server_callbacks.wait() returns as the number of closed matches // unit test exits
1. Thread 1: close_callback()  // oops!

Reply:
I see, good call.  Will fix

Reply:
done.  thank you @silentdai for the help